Added autograd support for torch.orgqr

IvanYashchuk opened the pull request:
This PR adds autograd support for `torch.orgqr`.

Since `torch.orgqr` is one of few functions that expose LAPACK's naming and all other linear algebra routines were renamed a long time ago, I also added a new function with a new name and `torch.orgqr` now is an alias for it.

The new proposed name is `householder_product`. For a matrix `input` and a vector `tau` LAPACK's orgqr operation takes columns of `input` (called Householder vectors or elementary reflectors) scalars of `tau` that together represent Householder matrices and then the product of these matrices is computed. See https://www.netlib.org/lapack/lug/node128.html.
Other linear algebra libraries that I'm aware of do not expose this LAPACK function, so there is some freedom in naming it. It is usually used internally only for QR decomposition, but can be useful for deep learning tasks now when it supports differentiation.

Resolves #50104

IvanYashchuk committed:
Make sure that contiguous tau is passed to lapack

IvanYashchuk committed:
Implemented orgqr_backward

IvanYashchuk committed:
Updated comments in orgqr_backward

IvanYashchuk committed:
Added support for complex differentiation

IvanYashchuk committed:
Added OpInfo for orgqr

IvanYashchuk committed:
Rename self -> input; input2 -> tau

IvanYashchuk committed:
Renamed orgqr -> householder_product

orgqr is now an alias of householder_product
added documentation for householder_product

IvanYashchuk committed:
Modified backward to work for rectangular inputs, added sample inputs for OpInfo

IvanYashchuk commented:
Probably complex gradgradcheck should be marked as slow.
```
============================================================ slowest 10 durations =============================================================
11.94s call     test/test_ops.py::TestGradientsCPU::test_fn_gradgrad_householder_product_cpu_complex128
3.36s call     test/test_ops.py::TestGradientsCPU::test_fn_gradgrad_householder_product_cpu_float64
0.96s call     test/test_ops.py::TestOpInfoCUDA::test_supported_dtypes_householder_product_cuda_complex128
0.36s call     test/test_ops.py::TestGradientsCPU::test_fn_grad_householder_product_cpu_complex128
0.15s call     test/test_ops.py::TestGradientsCPU::test_fn_grad_householder_product_cpu_float64
0.13s call     test/test_ops.py::TestCommonCPU::test_variant_consistency_jit_householder_product_cpu_complex64
0.13s call     test/test_ops.py::TestCommonCPU::test_variant_consistency_jit_householder_product_cpu_complex128
0.12s call     test/test_ops.py::TestCommonCPU::test_variant_consistency_jit_householder_product_cpu_float32
0.12s call     test/test_ops.py::TestCommonCPU::test_variant_consistency_jit_householder_product_cpu_float64
0.09s call     test/test_ops.py::TestOpInfoCUDA::test_unsupported_dtypes_householder_product_cuda_int32
```

IvanYashchuk committed:
Merge remote-tracking branch 'upstream/master' into orgqr-backward

IvanYashchuk committed:
Fixed overrides

mruberry commented on aten/src/ATen/native/BatchLinearAlgebra.cpp:
```diff
@@ -1290,15 +1297,24 @@ Tensor& orgqr_out_info(const Tensor& input, const Tensor& tau, Tensor& result, T
   TORCH_CHECK(input.size(-2) >= input.size(-1), "orgqr: input.shape[-2] must be greater than or equal to input.shape[-1]");
   TORCH_CHECK(input.size(-1) >= tau.size(-1), "orgqr: input.shape[-1] must be greater than or equal to tau.shape[-1]");
 
+  TORCH_CHECK(input.dim() - tau.dim() == 1, "orgqr: Expected tau to have one dimension less than input, but got tau.ndim equal to ",
```
Wouldn't the names for these operations start with "householder_product:" now? Or maybe even "torch.linalg.householder_product():"?

IvanYashchuk replied:
Yes, I should change that.

mruberry commented on aten/src/ATen/native/BatchLinearAlgebra.cpp:
```diff
@@ -1351,12 +1367,20 @@ Tensor& orgqr_out(const Tensor& input, const Tensor& tau, Tensor& result) {
   return result;
 }
 
+Tensor& orgqr_out(const Tensor& input, const Tensor& tau, Tensor& result) {
```
Add a comment her explaining this alias

mruberry commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -6000,14 +6000,20 @@
-  dispatch:
-    CPU: orgqr
+
+- func: orgqr.out(Tensor self, Tensor input2, *, Tensor(a!) out) -> Tensor(a!)
```
So orgqr was always CPU-only and this is also CPU-only for now?

IvanYashchuk replied:
Yes, it was always CPU-only. CUDA support is in https://github.com/pytorch/pytorch/pull/51348 which now passes all CI checks and can be merged. After https://github.com/pytorch/pytorch/pull/51348 is merged I'll update this PR.

mruberry commented on torch/_torch_docs.py:
```diff
@@ -6628,21 +6686,18 @@ def merge_dicts(*dicts):
-See `LAPACK documentation for orgqr`_ for further details.
+This directly calls the underlying LAPACK function `orgqr`.
+
+Alias for :func:`torch.householder_product`.
```
Maybe we should just have the "Alias for ..." line? The rest is redundant with it.

mruberry commented:
ASAN failure looks real.

mruberry commented on torch/csrc/autograd/FunctionsManual.cpp:
```diff
@@ -2987,6 +2987,105 @@ bool any_variable_defined(variable_list& variables) {
+    return std::tuple<Tensor, Tensor>(Tensor(), Tensor());
+  }
+
+  // LAPACK's orgqr operation computes a product of Householder matrices.
```
Excellent work on the comments here.

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -1005,6 +1005,35 @@ def sample_inputs(self, device, dtype, requires_grad=False):
+        SampleInput((make_tensor((0, 0), device, dtype, low=None, high=None, requires_grad=requires_grad),
+                    make_tensor((0,), device, dtype, low=None, high=None, requires_grad=requires_grad))),
+
+        SampleInput((make_tensor((S, S), device, dtype, low=-2, high=2, requires_grad=requires_grad),
```
Would you elaborate on the range here?

IvanYashchuk replied:
Each column of the matrix is getting multiplied many times leading to very large values for the Jacobian matrix entries and making the finite-difference result of grad check less accurate. That's why gradcheck with the default range [-9, 9] fails.

mruberry replied:
Cool; would you add a comment explaining that?

mruberry commented on torch/_torch_docs.py:
```diff
@@ -3672,6 +3672,64 @@ def merge_dicts(*dicts):
+to explitly form the `Q` matrix of the QR decomposition.
+See `Representation of Orthogonal or Unitary Matrices`_ for further details.
+
+.. note:: LAPACK's `orgqr` is used for the computations.
```
Let's be explicit this only supports CPU inputs.

IvanYashchuk replied:
Now it supports GPU as well.

mruberry commented on torch/_torch_docs.py:
```diff
@@ -3672,6 +3672,64 @@ def merge_dicts(*dicts):
+:attr:`tau` is a tensor of scalar scale factors for each Householder vector.
+For the single matrix :attr:`input` taking its `i`-th column as `vᵢ` and `τ = tau[i]`
+gives the `i`-th Householder matrix as `Hᵢ = I − τ vᵢ vᵢᴴ`.
+The result of this function is `H₁ H₂ ... Hᵣ`, where `r` is equal to `tau.shape[-1]`.
```
New paragraph here

mruberry commented on torch/_torch_docs.py:
```diff
@@ -3672,6 +3672,64 @@ def merge_dicts(*dicts):
+Householder matrices are stored in compressed vector form as columns of a `(m × n)` matrix :attr:`input`,
+or columns of each matrix in a batched :attr:`input`.
+:attr:`tau` is a tensor of scalar scale factors for each Householder vector.
+For the single matrix :attr:`input` taking its `i`-th column as `vᵢ` and `τ = tau[i]`
```
What should the shape of tau be?

mruberry commented on torch/_torch_docs.py:
```diff
@@ -3672,6 +3672,64 @@ def merge_dicts(*dicts):
 
+add_docstr(torch.householder_product,
+           r"""
+householder_product(input, tau, *, out=None) -> Tensor
```
Should this be torch.linalg.householder_product()?

mruberry reviewed (commented):
Hey @IvanYashchuk! Overall this PR looks very good and I'm sorry you've been waiting for this review.

I made a few inline comments. My biggest question is: should this be in the torch.linalg namespace? 

Also it looks like one of the test failures is related to the PR. Looking forward to an update here!

IvanYashchuk commented:
> should this be in the torch.linalg namespace?

Currently `torch.linalg` contains only `numpy.linalg` compatible functions. I think it was not decided yet whether all linear algebra related functions (the ones that take matrices or batch of matrices as input) should live under `torch.linalg`. If we decide that it should be that way then we can immediately put this function under linalg namespace. It should be consistent and all other functions should be moved then as well for 1.9 release possibly with SciPy compatibility.

mruberry commented:
> > should this be in the torch.linalg namespace?
> 
> Currently `torch.linalg` contains only `numpy.linalg` compatible functions. I think it was not decided yet whether all linear algebra related functions (the ones that take matrices or batch of matrices as input) should live under `torch.linalg`. If we decide that it should be that way then we can immediately put this function under linalg namespace. It should be consistent and all other functions should be moved then as well for 1.9 release possibly with SciPy compatibility.

I think you're part of that decision as a torch.linalg lead ;)

Let me add some context to help inform our decision:

- A frequent complaint I hear is that numpy.linalg vs. numpy makes no sense. So it'd be nice if we had some principles determining when something should be in torch.linalg vs torch. Conceptually it'd be nice if all "linear algebra" operations were just in torch.linalg. Where this runs into some trouble is that very common operations, like matmuls, might be difficult to actually deprecate and move, but maybe a good first rule would be "all new linear algebra operations go in torch.linalg."
- @kurtamohler already helped us design torch.linalg.vector_norm and torch.linalg.matrix_norm, so the namespace will already contain operations not in NumPy or SciPy's linalg namespaces. I think this is a cool feature we'll want to talk about in our 1.9 blog post (that is, how we're going beyond in our design of the linalg module).
- We'll also probably have new features in our linalg module, like those coming from the historic PyTorch linalg functions and  @Balandat's idea for being able to skip error syncs on CUDA devices.

We don't have to decide this immediately. In fact we should probably schedule another torch.linalg sync. We could talk about it for a bit then, too, if you like.

nikitaved commented:
@IvanYashchuk , sorry for the delay. I will check the math soon!

IvanYashchuk committed:
Merge remote-tracking branch 'upstream/master' into orgqr-backward

IvanYashchuk committed:
Moved householder_product to linalg module

IvanYashchuk committed:
Trying to fix asan error in householder_product_backward

IvanYashchuk committed:
Merge remote-tracking branch 'upstream/master' into orgqr-backward

IvanYashchuk committed:
Enable cuda opinfo tests for householder_product

IvanYashchuk committed:
Fixed cuda for tau.shape[-1] == 0

IvanYashchuk committed:
Trailing space

IvanYashchuk committed:
Remove unused import

IvanYashchuk committed:
Merge remote-tracking branch 'upstream/master' into orgqr-backward

IvanYashchuk commented:
@mruberry I moved `householder_product` to the `linalg` module.
I also updated this PR so that CUDA is also supported with https://github.com/pytorch/pytorch/pull/51348. Gradchecks tests revealed that cuSOLVER doesn't return the correct result for the empty `tau` argument, which is fixed in https://github.com/pytorch/pytorch/pull/52637/commits/3a99d1f3c7d6643663e1b00f45290811c26e2f40.

mruberry commented:
The ASAN failure looks real, unfortunately.

IvanYashchuk committed:
Pass indices with {}; should fix asan error

IvanYashchuk committed:
Merge remote-tracking branch 'upstream/master' into orgqr-backward

nikitaved commented on torch/csrc/autograd/FunctionsManual.cpp:
```diff
@@ -3097,6 +3097,105 @@ bool any_variable_defined(variable_list& variables) {
+
+    // we need to recompute input[j] * at::outer(v, v)
+    auto tau_unsqueezed = tau.index({"...", j}).unsqueeze(-1);  // tau[..., j][:, None]
+    auto tau_unsqueezed2 = tau_unsqueezed.unsqueeze(-1);  // tau[..., j][:, None, None]
+    auto v3 = at::matmul(v1.unsqueeze(-1), v2.conj().unsqueeze(-2));  // batch outer product, at::outer doesn't work for batched input
+    auto v4 = tau_unsqueezed2 * v3;
```
`v3` can be computed as ` v.unsqueeze(-1) * v`.

IvanYashchuk replied:
Yes, `v.unsqueeze(-1) * v.conj().unsqueeze(-2)` will work.

nikitaved commented on torch/csrc/autograd/FunctionsManual.cpp:
```diff
@@ -3097,6 +3097,105 @@ bool any_variable_defined(variable_list& variables) {
+  d_result.index_put_({"...", Slice(), Slice(None, input.size(-1))}, grad); // d_result[..., :, :input.shape[-1]] = grad
+
+  auto start_j = tau.size(-1) - 1;
+  for (int64_t j = start_j; j >= 0; j--) {
+    auto v = input_.index({"...", Slice(), j});
+    auto v1 = v, v2 = v;
```
I believe you can replace indexing with ellipsis with `narrow`, should be faster.

IvanYashchuk replied:
In my tests, replacing `.index` here and later in code with `.narrow` doesn't make any difference.

nikitaved replied:
Of course, compared to matrix products it is almost nothing, so, I guess, it is primarily a taste thingy. I have just never seen `index` used in ATen, so good to know this is yet another tool available :)

nikitaved commented:
@mruberry , in our discussion with @IvanYashchuk  we came to the conclusion that his PR could be merged as of now.
However, there an efficient algorithm for this backward which is simpler, but requires some theoretical insight and basic properties of reflectors. The cool part of this algorithm is that it requires only a single matrix-matrix product, the rest could be done with matrix-vector routines, which results in total complexity of 2-4 matrix products. I could submit a PR later.

IvanYashchuk committed:
matmul is not needed for the outer product

IvanYashchuk committed:
Merge remote-tracking branch 'upstream/master' into orgqr-backward

IvanYashchuk committed:
Added a comment explaining non-default range for sample_inputs_householder_product

IvanYashchuk committed:
Added a comment that orgqr is an alias of householder_product

IvanYashchuk committed:
Added householder_product to linalg.rst

IvanYashchuk commented:
@mruberry ASAN failure is fixed. CI failures are not related. It's ready for another round of review and probably can be merged.

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -367,6 +367,64 @@
+to explitly form the `Q` matrix of the QR decomposition.
+See `Representation of Orthogonal or Unitary Matrices`_ for further details.
+
+.. note:: LAPACK's `orgqr` is used for the computations.
```
What about on CUDA?

IvanYashchuk replied:
Added a note: "For CUDA inputs, cuSOLVER's `orgqr` routine is used if CUDA version >= 10.1.243."

mruberry reviewed (approved):
Thank you for helping review this @nikitaved and providing an update. And great work to both you and @IvanYashchuk!

It's exciting to have new linalg functions in torch.linalg!

This looks good but you would rebase past the doc failures? I'm also curious what's going on the XLA build. I made one tiny docs note about explaining what's called on CUDA (since orgqr now supports CUDA) that we might as well resolve during the rebase.

IvanYashchuk committed:
Merge remote-tracking branch 'upstream/master' into orgqr-backward

IvanYashchuk committed:
Added a note about cusolver being used for cuda inputs

IvanYashchuk commented:
@mruberry I updated the branch and added a note that cuSOLVER is used for CUDA inputs. Let's see whether doc and xla CI failures go away.
